### PR TITLE
Allow the user to edit experiment information via dialog 

### PR DESCRIPTION
**Pull request by Steffengreiner**
**What was changed**
1.) Replacement of the ExperimentCreationDialog with `ExperimentInformationDialog` to allow prefilling for editing purposes within an Experiment similiar to `ExperimentalVariableDialog`
2.) Update `ExperimentSupportComponent` and `ExperimentListComponent` to employ the`Context` for retrieval of ExperimentIds  and ProjectIds
3.) Enable the user to edit experiment information in the `ExperimentDetailsComponent` via the `ExperimentInformationDialog`
4.) A change of the experiment Information is propagated to the `ExperimentMainComponent `via Listeners, which will trigger a reload of the page to show the changes also in the Experiment List

### CODE REVIEW

File: projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/experiment/Experiment.java [hunk 1]
```diff
@@ -18,6 +20,7 @@
 import life.qbic.projectmanagement.domain.project.experiment.vocabulary.Analyte;
 import life.qbic.projectmanagement.domain.project.experiment.vocabulary.Species;
 import life.qbic.projectmanagement.domain.project.experiment.vocabulary.Specimen;
+import org.springframework.util.CollectionUtils;
```
Comment (KochTobi) [inline projectmanagement/src/main/java/life/qbic/projectmanagement/domain/project/experiment/Experiment.java:23]: Why not use java functions like isEmpty or Objects.isNull here?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java [hunk 2]
```diff
@@ -129,20 +139,19 @@ private DisclaimerCard createNoVariableDisclaimer() {
   }
 
   private void layoutComponent() {
+    this.add(header);
+    header.addClassName("header");
     this.add(content);
+    //Necessary to avoid css collution
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java:145]: ?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java [hunk 3]
```diff
@@ -152,6 +161,39 @@ private void configureComponent() {
...
+          editDialog.addCancelEventListener(
+              experimentInformationDialogCancelEvent -> editDialog.close());
+          editDialog.addConfirmEventListener(experimentInformationDialogConfirmEvent -> {
+            ExperimentInformationContent experimentInformationContent = experimentInformationDialogConfirmEvent.getSource()
+                .content();
+            experimentInformationService.editExperimentInformation(experimentId,
+                experimentInformationContent.experimentName(), experimentInformationContent.species(),
+                experimentInformationContent.specimen(), experimentInformationContent.analytes());
+            editDialog.close();
+            fireEditEvent();
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java:187]: please try to avoid multi-line lambda expressions! Try to extract meaningful methods instead

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java [hunk 4]
```diff
@@ -407,10 +466,10 @@ private void loadExperimentInfo(Experiment experiment) {
     this.experimentSummary.removeAll();
     this.experimentSummary.add(factSheet);
     factSheet.showMenu();
-    reloadExperimentalVariables(experiment);
+    reloadExperimentInformation(experiment);
   }
 
-  private void reloadExperimentalVariables(Experiment experiment) {
+  private void reloadExperimentInformation(Experiment experiment) {
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java:472]: why rename the method? as far as I can see this still only reloads the variables

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentInformationDialog.java [hunk 5]
```diff
@@ -0,0 +1,166 @@
...
+
+  private static ExperimentInformationDialog editDialog(
+      ExperimentalDesignSearchService experimentalDesignSearchService, String experimentName,
+      Collection<Species> species, Collection<Specimen> specimen, Collection<Analyte> analytes) {
+    ExperimentInformationDialog experimentInformationDialog = new ExperimentInformationDialog(
+        experimentalDesignSearchService, true);
+    experimentInformationDialog.setExperimentInformation(experimentName, species, specimen,
+        analytes);
+    return experimentInformationDialog;
+  }
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentInformationDialog.java:97]: what is the difference between prefilled and editDialog?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java [hunk 6]
```diff
@@ -75,7 +74,7 @@ public ExperimentInformationMain(
     this.projectInformationService = projectInformationService;
     this.experimentInformationService = experimentInformationService;
     layoutComponent();
-
+    addListeners();
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:77]: Why move this to the constructor?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java [hunk 7]
```diff
@@ -138,20 +136,18 @@ private ExperimentId activeExperiment(ProjectId parsedProjectId) {
 
   private void setContext(Context context) {
     experimentContentComponent.setContext(context);
-    experimentSupportComponent.projectId(context.projectId().orElseThrow());
-    experimentSupportComponent.setExperiments(
-        experimentInformationService.findAllForProject(context.projectId().orElseThrow()));
+    experimentSupportComponent.setContext(context);
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:139]: What happened here?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java [hunk 8]
```diff
@@ -138,20 +136,18 @@ private ExperimentId activeExperiment(ProjectId parsedProjectId) {
...
     this.context = context;
   }
-
-
   private void addListeners() {
     experimentSupportComponent.addExperimentSelectionListener(
         event -> routeToExperiment(event.getSource().experimentId()));
     experimentSupportComponent.addExperimentCreationListener(
         event -> routeToExperiment(event.experimentId()));
+    experimentContentComponent.addExperimentEditListener(
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:149]: Why reroute? We are already at this experiment. This leads to an additional loading of the experiment.

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java [hunk 9]
```diff
@@ -60,68 +62,69 @@ public ExperimentListComponent(
     Objects.requireNonNull(addExperimentToProjectService);
     Objects.requireNonNull(experimentalDesignSearchService);
     this.addExperimentToProjectService = addExperimentToProjectService;
-    this.experimentCreationDialog = new ExperimentCreationDialog(experimentalDesignSearchService);
+    this.experimentInformationService = experimentInformationService;
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java:65]: What happened in this class?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java [hunk 10]
```diff
@@ -66,9 +62,8 @@ public static ExperimentItemCollection create(String labelCreationCard) {
    */
   public void addExperimentItem(ExperimentItem item) {
     items.add(item);
-    add(item);
-    addComponentAsLast(createExperiment);
-    subscribeToClickEvent(item);
+    addComponentAsFirst(item);
+    addSelectionEventListener(item);
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java:66]: why did you change the behaviour here? from addin it in the end of the list to adding it on top? (not part of this PR I would say)

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java [hunk 11]
```diff
@@ -136,10 +131,8 @@ public void addCreateEventListener(ComponentEventListener<CreationClickedEvent>
   @Override
   public void removeAll() {
     super.removeAll();
-    addCreationCard(createExperiment);
+    addComponentAsLast(createExperiment);
```
Comment (KochTobi) [inline vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java:134]: why did you remove the listener deletion?

**Review by KochTobi — CHANGES_REQUESTED** [review]
Hi Steffen,
Nice changes. I have a lot of questions

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java (hunk 4 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java:472]: Oh good point I changed it back 👍

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentInformationDialog.java (hunk 5 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentInformationDialog.java:97]: `prefilled()` is the public accessor method to be used if one wants to create an ExperimentalVariablesDialog in edit mode. 
I tried to adhere to the structure introduced in the `ExperimentalVariablesDialog` (see screenshot below) so the dialogues can be created in  the same way. 🤔 
Bildschirmfoto 2023-08-08 um 10 20 45

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java (hunk 7 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:139]: I introduced the Context object holding the context information in the ExperimentInformationMain and propagate it to the supportcomponent instead of the experimentID and projectID explicitly

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java (hunk 8 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:149]: Yes and that is intended so the ExperimentListComponent also shows the changed experimentinformation.

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java (hunk 6 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:77]: The previous behaviour was to add the listeners everytime a beforeEnterEvent happens, which is unnecessary if the listeners can be instantianted once in the constructor with the same behaviour.

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java (hunk 9 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java:65]: Previously, we propagated the `ExperimentInformationService` to this class but never used it.
Therefore I changed the logic in this class so it only receives the `Context` Information from the Supportcomponent instead of a `Collection<Experiment>` 
The `ExperimentListComponent` then retrieves all the experiment of the provided projectId in the `Context` via the `ExperimentInformationService`
Finally the `ExperimentInformationDialog` replaces the` ProjectCreationDialog` and is not stored as a class variable but built anew every time, similiar to the `ExperimentalVariableDialog`

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java (hunk 11 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java:134]: It's unclear to me why we need to remove the listeners. 
Especially since removing the createListener robs the component of it's functionality. 🤔

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java (hunk 3 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java:187]: That's a good point 👍

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java (hunk 8 above)
Comment (KochTobi) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:149]: Then I would argue that the parent (main component) should listen to context changes and propagate those to the support section.

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java (hunk 9 above)
Comment (KochTobi) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java:65]: Do we want to fetch all experiments in the component or in the main?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java (hunk 10 above)
Comment (KochTobi) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/support/experiment/ExperimentItemCollection.java:66]: was this not already the case?

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java (hunk 2 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/experiment/ExperimentDetailsComponent.java:146]: As we discussed there is an issue that the css from the .content class of the experiment-details-component interacts with the .content css class of the experimental-groups-collection and the .content of the experimental-variable. Since this seems to be a bigger issue i proposed a small fix for this PR

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java (hunk 9 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentListComponent.java:65]: I think it makes more sense to fetch the experiments in the list component since only there they are needed

File: vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java (hunk 8 above)
Comment (Steffengreiner) [reply vaadinfrontend/src/main/java/life/qbic/datamanager/views/projects/project/experiments/ExperimentInformationMain.java:149]: Alright I added a dedicated listener within the experimentdetailscomponent and propagate the new context to the support component once the edit event was successful

**Review by KochTobi — APPROVED** [review]
